x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/exchange/ExchangeService.java

```diff
@@ -361,6 +353,15 @@ private void doFetchPageAsync(boolean allSourcesFinished, ActionListener<Exchang
                 }, responseExecutor)
             );
         }
+
+        @Override
+        public void close(ActionListener<Void> listener) {
+            if (finished.compareAndSet(false, true)) {
+                doFetchPageAsync(true, listener.delegateFailure((l, unused) -> l.onResponse(null)));
```

Comment:
what happens here if the remote goes away or is slow? Does the whole request wait for the closing request to complete, and will it fail if it fails?

Reply:
The entire request must wait for the closing. The (implicit) rule for the task is that once it is completed, its child tasks should also be completed, and its resources should be cleaned up unless that's not possible. Here, we only perform the actual `closing()` when encountering exceptions; in other cases, the remote sink should finish on its own.